src/ast/nodes/shared/FunctionBase.ts

```diff
@@ -27,6 +34,18 @@ import {
 import type { ObjectEntity } from './ObjectEntity';
 import type { PatternNode } from './Pattern';
 
+type FunctionParameterState =
```

Comment:
I found this TOP/MID/BOTTOM terminology quite confusing. After studying the code, I understand that TOP means "no value yet", MID means "on or several values that are either the same expression or the same known literal value" and BOTTOM means "several different values". Maybe we could improve readability here in one way or another and avoid the arbitrary terminology, maybe even avoid the recreated ad-hoc objects?

First question: Do we really need to initialize the array, or couldn't a value of `undefined` or in general a falsy value carry the same meaning?
Then instead of MID, we just put the expressions themselves into the array, and instead of BOTTOM either put UNKNOWN_EXPRESSION (which is technically also an expression, but of course would have the correct effect on the parameter) or a Symbol value with a helpful name.

Reply:
Sorry for lacking the context, I learned these TOP/MID/BOTTOM terminology from constant folding like [here](http://infolab.stanford.edu/~ullman/dragon/w06/lectures/cp.pdf). Sure for readability and performance we should improve the naming and usage.

1. No, we don't need to initialize the array, the default `undefined` is enough
2. Using Symbol for BOTTOM is a good idea 👍since `UNKNOWN_EXPRESSION` is an `ExpressionEntity`, not a node with `parent`, for clarity I would like to create `const UnknownArgument = Symbol('Unknown Argument')

Then: `type FunctionParameterState = ExpressionNode | SpreadElement | typeof UnknownArgument